### PR TITLE
fix(groovy) strings are not allowed inside ternary clauses

### LINKED ISSUE #2217: (groovy) strings are not allowed inside ternary clauses
If you do for example:

```groovy
def createCodeStyleMsgBody(build, label) {
    def codeStyleFiles = findCodestyleFilesInLog(build)
    def formattingMsg = label < 0 ? ('The following files need formatting:\n    ' +
        codeStyleFiles.join('\n    ')) : 'All files are correctly formatted'
    def url = build.url + "consoleText"

    return "${Globals.resTicks[build.result]} $formattingMsg\n    (${url})"
}
```

'The following files need formatting:\n    ' is not highlighted correctly as a string, secondly everything after ' + is coloured in blue (so if you create another def after, it colours all that in blue).

### PR DESCRIPTION
**Pull request by yyyc514**
Closes #2217

### CODE REVIEW

**Comment by yyyc514** [comment]
Sorry, I should have split the commit in two.  Mostly it's just a reorg of the modes into constants to make them easier to reuse inside TERNARY.

File: src/languages/groovy.js [hunk 1]
```diff
@@ -77,25 +92,35 @@ export default function(hljs) {
...
+                REGEXP,
+                NUMBER,
+                'self'
+              ]
             },
             {
                 // highlight labeled statements
-                className: 'symbol', begin: '^\\s*[A-Za-z0-9_$]+:',
+                className: 'symbol',
+                begin: '^[ ]*' + regex.lookahead(IDENT_RE + ':'),
```
Comment (egor-rogov) [inline src/languages/groovy.js:118]: You've changed \s to space, is it ok?

File: src/languages/groovy.js [hunk 2]
```diff
@@ -5,69 +5,84 @@
...
+    hljs.QUOTE_STRING_MODE,
+    ],
+    { className: "string" }
+  );
+
     return {
         name: 'Groovy',
         keywords: {
-            literal : 'true false null',
+            built_in: 'this super',
```
Comment (egor-rogov) [inline src/languages/groovy.js:69]: What is the reason to make these built-ins? I looked into https://docs.groovy-lang.org/latest/html/documentation/ and found them among keywords.

File: src/languages/groovy.js [hunk 3]
```diff
@@ -77,25 +92,35 @@ export default function(hljs) {
...
             },
-            hljs.C_NUMBER_MODE,
             {
                 className: 'meta', begin: '@[A-Za-z]+'
             },
             {
-                // highlight map keys and named parameters as strings
-                className: 'string', begin: /[^\?]{0}[A-Za-z0-9_$]+ *:/
+              // highlight map keys and named parameters as attrs
+              className: 'attr', begin: IDENT_RE + '[ ]*:'
```
Comment (egor-rogov) [inline src/languages/groovy.js:100]: Should here be \s in place of a space?..

File: src/languages/groovy.js (hunk 1 above)
Comment (yyyc514) [reply src/languages/groovy.js:118]: This is so the match starts on the current line, not many lines previous.

File: src/languages/groovy.js (hunk 2 above)
Comment (yyyc514) [reply src/languages/groovy.js:69]: They are still keywords, everything in `keywords` is keywords, the question is what kind of keyword on they.  These are very clearly not language keywords in the same vein as `for`, `if`, etc... they are built in variables, same as `global` or `window`, etc...

File: src/languages/groovy.js (hunk 3 above)
Comment (yyyc514) [reply src/languages/groovy.js:100]: See prior note, again do not want to eat new lines.

**Comment by yyyc514** [comment]
>  I mean that \t is also a possible character.

Ugh, I guess... if I have to worry about that everywhere I'm going to go and write some kind of whiteSpaceButNotNewline helper. :-)

File: src/languages/groovy.js (hunk 1 above)
Comment (yyyc514) [reply src/languages/groovy.js]: Adding support for tabs.

File: src/languages/groovy.js (hunk 3 above)
Comment (yyyc514) [reply src/languages/groovy.js]: Adding support for tabs.

**Pull request merged**